### ISSUE TITLE
Code review

### ISSUE THREAD

**Issue by abuelhassan** [akorwash/QuizBattle#2]
* You should add `.vscode` file to `.gitignore`.
If you are working within a team, you don't want to enforce your editor's settings on other team members. Or even to others who want to contribute to your repo.

* I see some inconsistencies in your naming for folders and files. Some are camelCase and others are SnakeCase.
I think the convention is to use lower case letters for both folders and files. Here are a few examples from `net/http`
`filetransport.go`, `httputil`

* Check these slides for some naming tips.
https://talks.golang.org/2014/names.slide#1

* Avoid nesting if possible. For example:
```
func (startup *Startup) LoadQuestions() *Startup {
	if startup.Errors != nil {
		return startup
	}

	if len(engine.QuestionSet) <= 0 {
		for i := 1; i <= 100; i++ {
			question := engine.NewQuestion(i, "test question #"+strconv.Itoa(i))
			engine.QuestionSet = append(engine.QuestionSet, *question)
		}
	}
	return startup
}
```
Better be:
```
func (startup *Startup) LoadQuestions() { 
        const lim = 100 // use constants instead of the magic values and choose a better name than `lim` I don't know the context so I couldn't choose a better name 😢 . You can also decide if this needs to go global, be exported, add it to Startup struct, etc...

	if startup.Errors != nil || len(engine.QuestionSet) == 0 {
		return
	}
	for i := 1; i <= lim; i++ {
		q := engine.NewQuestion(i, "test question #"+strconv.Itoa(i))
		engine.QuestionSet = append(engine.QuestionSet, *q)
	}
}
```

* In datastore/entities instead of having multiple files, I think it's better to have them all in `datastore/entities/entities.go`
And as the package name is called `entities` so you better drop the `EntitySuffix` from the structs' names.
Instead of `type AnswerEntity struct{}`, let it be `type Answer struct{}`
Put in mind that when you use this from outside the package you will write `entities.AnswerEntity` you will find that entity is written twice, while entities. `entities.Answer` is clean enough.

* I see you are already applying some good practices, like sorting and grouping the imports, and others 👍 Good work 🎉 .

* There are other parts with areas of improvement. But these are the points I noticed from a glance.
If you find it helpful, I will try to give more feedback when I get some free time.

* ~Pro~ Bro tips:
  - Check this [youtube channel](https://www.youtube.com/channel/UC_BzFbxG2za3bp5NRRRXJSw), there are a couple of videos where the guy clones a repo, and refactor it. I have learned a lot from this guy.
  - Explore some of the standard libraries. They have great stuff, and I have learned a couple of useful patterns while checking them, you can also get some good practices from them.
  - Write more documentation, maybe update the README file with some instructions on how to run the project locally.
Also check [this](https://golang.org/doc/effective_go.html#commentary)
  - Dude, you are missing a lot of fun by not writing unit tests!
There's a couple of plugins to help you generate a skeleton for testing a function. They almost do all the work, and you will have to only fill a table by some data/structs.
Give a try to writing unit tests and even better if you work in TDD, like solid programmers (I am not one of them yet, ed3ily).
A very popular way is `Table Driven Tests`. You can check it on some of the standard libraries like `net/http`.

* A recommendation:
From the parts I have checked, I didn't see any usage of channels. Which is fine, but if you want a simple task to use/learn channels, check how you can do a graceful shutdown, so that when the user terminates the program by ctrl+c, the program stop taking any actions from the user but keeps the system up till it shutdowns all running services successfully. Or maybe to save the state of your game or whatever is in `engine.ExitTheGame()`. 
And maybe rename `ExitTheGame` to `Shutdown`, `Exit` or `Close`.

**Comment by akorwash** [comment]
thanks a lot Mohamed,
Now i working to fix all of this issue and already fix most of them.
You can see I start use Gorilla Mux so i will consider all of you mentioned in this code review.

I hope you find some time to have another look when i convert the code to APIs.
after this i will try to write unite tests for the project and start another journey with CI CD.

**Comment by akorwash** [comment]
I have a question when i try to deploy my project to heroku using CI CD
i can't deploy it because on GitHub my engine folder start with E capital "Engine" however i found it as "engine" in visual studio code?

also why i have to name my module as my GitHub link "github.com/akorwash/QuizBattle"?

**Comment by abuelhassan** [comment]
For the Heroku issue, it may be the difference between OSes. 
On Linux, the file system is case sensitive, while on Windows it's case insensitive. I am not sure if that is your case, I am just guessing. 
Anyways, I prefer having files and folders named in lowercase.

**Comment by abuelhassan** [comment]
You can check the error messages convention from here: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

I also like this new way of handling errors, didn't have the chance to work with it yet, but it looks good.
https://blog.golang.org/go1.13-errors

**Comment by akorwash** [comment]
i will check the new way also, i think it's something like catch multiple exception at .NET.
so i think i am familiar with this concept.

**Comment by akorwash** [comment]
@abuelhassan Now i am added some unit tests
if you have some time to check if it and give me your feedback.

thanks in advance

**Comment by MohamedAmrAbuelhassan** [comment]
I will give it a look later today :+1:

Just a quick note:
In Go, it's a common practice to have the test files for **Unit tests,** in the same folders with code.
check (this)[https://github.com/golang/go/tree/master/src/net/http] from the standard library.
```
- http
  -- http.go
  -- http_test.go
```
This way enables you to create unit tests for the non-exported functions.

In the go src code, they use the `/test` directory for different types of testing, but not for unit testing. Check (this)[https://github.com/golang/go/tree/master/test]

Also for the unit tests, check the (table-driven-tests)[https://github.com/golang/go/wiki/TableDrivenTests] approach.
If you decided to follow the (table-driven-tests) approach for unit testing, check this (plugin)[https://github.com/cweill/gotests].
Also check (this video)[https://www.youtube.com/watch?v=hVFEV-ieeew], to know more about the Table Driven Tests. The video shows why and how to use it.

**Comment by abuelhassan** [comment]
I will give it a look later today :+1:

Just a quick note:
In Go, it's a common practice to have the test files for **unit tests,** in the same folders with code.
check [this](https://github.com/golang/go/tree/master/src/net/http) from the standard library.
```
- http
  -- http.go
  -- http_test.go
```
This way enables you to create unit tests for the non-exported functions. There may be other reasons, but this is the obvious one for me.

In the go src code, they use the `/test` directory for other types of testing, but not for unit testing. Check [this](https://github.com/golang/go/blob/master/test/README.md)

Also for the unit tests, check the [table-driven-tests](https://github.com/golang/go/wiki/TableDrivenTests) approach.
If you decided to follow the (table-driven-tests) approach for unit testing, check this [plugin](https://github.com/cweill/gotests).
Also check [this video](https://www.youtube.com/watch?v=hVFEV-ieeew), to know more about the Table Driven Tests. The video shows why and how to use it.

**Comment by akorwash** [comment]
I would love to use this
```
- http
  -- http.go
  -- http_test.go
```

also i will check how can i use TDD "table" maybe it much easier for me.

Thanks

**Comment by abuelhassan** [comment]
### About the code structure
First of all, a disclaimer, I am still learning how to come up with a good structure. So I am just saying an opinion, and I would like to know your opinion so that I learn something new.

I have read these two articles about clean architecture.
[Clean Architecture, the right way](https://medium.com./gdg-vit/clean-architecture-the-right-way-d83b81ecac6) and [The clean architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html)
(Lifehack, to make `medium.com` works in Egypt, put a dot `.` after the `com`  -> `medium.com./`, this seems stupid but it works for me 😂 🇪🇬)

I find this structure clear
Client -> Handlers/Controllers -> Use Cases/Business Logic -> Entities
> The *Dependency Rule* says that source code dependencies can only point inwards. Nothing in an inner circle can know anything at all about something in an outer circle.

So controllers know about the business logic layer, but not vice versa.

I find this part of the code, breaking the *Dependency Rule*. You shouldn't need to import a `handler` in a `service`.
https://github.com/akorwash/QuizBattle/blob/7003ed33dbf08b8b3cbc25ff123b448bc6d282e7/service/login/Iloginservice.go#L8
This may lead to circular dependencies. If you go to `startup.go` and put this line `_ = login.LoginFactory("amr", "123")` in your `Startup` function, you will see an error while building the project.
Even if you don't need the login in this handler, but it doesn't make sense to me.

<hr>

### Layers' responsibilities
For the responsibilities of each layer, here's what I prefer, it's my current understanding, which may differ in the future. So again, please share with me your opinion so that I learn something new.

#### In Handlers:
I always put these points in a handler.
1. I parse request(body, headers, query params, etc...) and convert it to input args for the `business logic` layer.
2. Call the `business logic` layer. (usually returns (Output, error))
3. `if err != nil` convert the error from the logic layer to another error with the appropriate message and status code.
4. Otherwise, I convert Output to the appropriate response (JSON, text, XML, etc...).

*Note*: I sometimes put caching in this layer, some other times I put it in the `business layer` depending on the case.
**But** I would move any validations to the `business logic` layer.

```go
// The reason why I have the errors in both the business layer and in the handler,
// that I think the business layer shouldn't know about the status code or the message client will receive.
// It also shouldn't care if the handler converts the output to Json, XML or any other foramt.
var (
	errNotFound   = buildError(http.StatusNotFound, "not found") // buildError may not be the best name.
	errEmptyID    = buildError(http.StatusBadRequest, "empty id")
	errBadRequest = buildError(http.StatusBadRequest, "bad request")
	errInternal   = buildError(http.StatusInternalServerError, "internal server error")
)


func GetArticle(w http.ResponseWriter, r *http.Request) {
	// logic to parse request
	...

	// call business logic layer
	out, err := article.New().Get(id) // No need to call a func GetByID if you don't get by other fields, also if it has `id` as a param already.
	if err != nil {
		switch err {
		case article.ErrEmptyID:
			sendError(errEmptyID)
		case article.ErrNotFound:
			sendError(errNotFound)
		...
		default:
			sendError(errInternal)
		}
		return
	}
	// logic to convert `out` to the appropriate form, and send it.
	...
}

// I prefer to list all handlers in the same file. handlers/handlers.go. Or at least under the same package.
```

I would also spend some time deciding what should be implemented in a handler, and what better be a middleware. It also depends on the case.
A middleware that I find useful, is a middleware to validate JWT tokens. Which parses the token to the user data and embed it in the request or sends an unauthenticated error if needed. 
I like this to be a middleware instead of implementing this step in every handler.

**Comment by akorwash** [comment]
I know about clean architecture i have been used it in different projects before using .NET Core,
I am senior .net developer also act as technical architect.

the big problem with me when i to use go, go have a different specs and convention and it's has it's own way to do things.
but when i find a project such as you send me using clean architecture i know what should i do to implement it using go.

i have been using c# for 10 years ago, it's easy for me to write any architecture clean/modern/onion  but go challenge me to apply this architecture.

go challenge me to  structure my project
where services, middlewares, repositories and entities should be and what the traditional convention should i follow.

when i am using c# and .net core applying onion architecture  i can create multiple libraries and import them easily,i can't do this is go but i think i am quick learner so i need to have a look for multiple projects and articles.

I see the both articles and now i will try to follow the same structure and restructure my project.

**Comment by abuelhassan** [comment]
Cool, I keep watching your repo, because I want to know what you will end up having. I am pretty sure I will learn a lot 👍 

The [medium](https://medium.com./gdg-vit/clean-architecture-the-right-way-d83b81ecac6) article I shared, is actually implemented in Go, that's why I shared it. But I honestly don't think that it covers much. Although it was a good start for me.

>the big problem with me when i to use go, go have a different specs and convention and it's has it's own way to do things.
but when i find a project such as you send me using clean architecture i know what should i do to implement it using go.

I feel you. I don't have much experience, so I am still learning how to come up with a clean code/architecture. And I know it's different in Golang than other languages.

Anyways, this architecture along with using interfaces and dependency injection properly, made it easy for me to write clean unit tests.
I used to say that I shouldn't modify code for the sake of writing easier unit tests. But then, I learned that the unit tests give me feedback about the code structure. So for example, if I am writing tests for a function and I am not able to mock a call from this function to another package, I would assume I am doing something wrong and I try to restructure some parts till it's easy to do this mocking.

Good luck 👍 Keep the good spirit up 🎉 .
